src/main/scala/org/apache/spark/shuffle/ucx/UcxShuffleTransport.scala

```diff
@@ -0,0 +1,138 @@
+/*
+* Copyright (c) 2020, NVIDIA CORPORATION. ALL RIGHTS RESERVED.
+* See file LICENSE for terms.
+*/
+package org.apache.spark.shuffle.ucx
+
+/**
+ * Class that represents some block in memory with it's address, size.
+ * @param isHostMemory - host or GPU memory
+ */
+case class UcxMemoryBlock(address: Long, size: Long, isHostMemory: Boolean = true)
+
+/**
+ * Opaque object to describe remote memory block (address, rkey, etc.).
+ */
+trait Cookie {
+  // Write this cookie to some address in memory
+  def writeToMemory(memory: UcxMemoryBlock)
+
+  // Size of this cookie in bytes
+  def size: Int
+}
+
+/**
+ * Base class to indicate some blockId. It should be hashable and could be constructed on both ends.
+ * E.g. ShuffleBlockId(shuffleId: Int, mapId: Long, reduceId: Int)
+ */
+trait BlockId
+
+/**
+ * Special case for metadata blocks, that always would be fetched using
+ * [[ UcxShuffleTransport.fetchBlocksByBlockIds() ]]. It has a method to get an actual blockId,
+ * it refers to,so when requested this MetadataBlockId, transport could know that soon blockId
+ * would be fetched, and may do some optimizations (pinning, pre-fetching, etc).
+ */
+trait MetadataBlockId extends BlockId {
+  def getBlockId: BlockId
+}
+
+trait Block {
+  // Transport will call this method when it would need an actual block memory.
```

Comment:
Should there be a boolean in this interface so that the transport knows this block's memory isn't changing (is available now)?

Reply:
Let's stick with the synchronous version for now. So when transport calls `block.getMemory` applications should:

- Block until other operation on this block is happening (compression, etc).
- Keep track of refCount. If transport requested memory for this block and not called yet `doneWithMemory` it means this block in progress now

Reply:
I wasn't referring to the synchronous behavior.

I was referring to the fact that some blocks may be ready to bake immediately (you could use this to pin the world as you want to do in some cases). It's really to help the cpu case.

Otherwise you have no way to know if the block can be pinned immediately or not (e.g. it's one of the fancy blocks that may need to be compressed, or some other expensive operation).

Reply:
Let's assume that calling `block.getMemory` always returns valid block memory at that point of time.  Transport *may* or *may not* pin it, or whatever. Now if you need to change blocks memory you can call `mutate` with a new block. Transport should finish all the outstanding requests with that block -> call `doneWithMemory` and use new block memory for new requests

Reply:
I think that's OK with us. The flag I had before was to try to help the CPU in the case where it would want a subset of blocks that should be pinned always.